### PR TITLE
fix: 1.sprite in world range 2.add scene mode 4. game "stepUnit" chan… 

### PR DESCRIPTION
**Pull request by sunqirui1987**
fix: 1.sprite in world range 2.add scene mode 4. game "stepUnit" change to mapconfig 5.window resize not change world map

### CODE REVIEW

File: game.go [hunk 1]
```diff
@@ -303,28 +300,42 @@ type costumeConfig struct {
 	Y                float64 `json:"y"`
 	FaceRight        float64 `json:"faceRight"` // turn face to right
 	BitmapResolution int     `json:"bitmapResolution"`
+	Mode             string  `json:"mode"`
```
Comment (xushiwei) [inline game.go:303]: costumeConfig is also used by sprite. so don't add `mode` in it.

File: game.go [hunk 2]
```diff
@@ -303,28 +300,42 @@ type costumeConfig struct {
...
+}
+
+func toSceneMode(mode string) int {
```
Comment (xushiwei) [inline game.go:306]: I think mapMode is better.

File: game.go [hunk 3]
```diff
@@ -303,28 +300,42 @@ type costumeConfig struct {
...
 type cameraConfig struct {
 	On string `json:"on"`
 }
 
 type mapConfig struct {
-	Mode   string `json:"mode"`
-	Width  int    `json:"width"`
-	Height int    `json:"height"`
+	Behavior string  `json:"behavior"`
+	StepUnit float64 `json:"stepUnit"`
```
Comment (xushiwei) [inline game.go:324]: maybe just call it `unit`? it's shared by `sprite.Move`, `sprite.Step` and `sprite.DistanceTo`, even `sprite.SetXpos`, `sprite.SetYpos`, `sprite.SetXYpos`, etc. (if possible), not only `sprite.Step`.

File: game.go [hunk 4]
```diff
@@ -303,28 +300,42 @@ type costumeConfig struct {
...
 
 type cameraConfig struct {
 	On string `json:"on"`
 }
 
 type mapConfig struct {
-	Mode   string `json:"mode"`
-	Width  int    `json:"width"`
-	Height int    `json:"height"`
+	Behavior string  `json:"behavior"`
```
Comment (xushiwei) [inline game.go:323]: we can only support one mode. so don't add `Behavior` at first.

File: game.go [hunk 5]
```diff
@@ -303,28 +300,42 @@ type costumeConfig struct {
...
-		return mapModeRepeat
-	case "fillCut":
-		return mapModeFillCut
-	case "fillRatio":
-		return mapModeFillRatio
+const (
+	mapInsideForBox = iota
+)
+
+func toMapInsideBehavior(behavior string) int {
```
Comment (xushiwei) [inline game.go:333]: add multi behaviour later. now we only support one.

File: spbase.go [hunk 6]
```diff
@@ -162,6 +162,7 @@ type costume struct {
 	img              delayloadImage
 	faceRight        float64
 	bitmapResolution int
+	mode             string
```
Comment (xushiwei) [inline spbase.go:165]: don't add mode to `costume`.

File: sprite.go [hunk 7]
```diff
@@ -695,14 +695,20 @@ func (p *Sprite) DistanceStepTo(obj interface{}) float64 {
 	x2, y2 := p.g.objectPos(obj)
```
Comment (xushiwei) [inline sprite.go:695]: remove `DistanceStepTo` api.

File: game.go (hunk 1 above)
Comment (sunqirui1987) [reply game.go:303]: so，game support one map mode ? I thought  each scene would have its own mode。

File: sprite.go [hunk 8]
```diff
@@ -746,12 +745,12 @@ func (p *Sprite) Step__2(step float64, animname string) {
 	if ani, ok := p.animations[animname]; ok {
 		anicopy := *ani
 		anicopy.From = 0
-		anicopy.To = step * float64(p.g.stepUnit)
+		anicopy.To = step * float64(p.g.gridUnit)
```
Comment (xushiwei) [inline sprite.go:748]: anicopy.To should be `step`. here we use `world axis`.

File: sprite.go [hunk 9]
```diff
@@ -746,12 +745,12 @@ func (p *Sprite) Step__2(step float64, animname string) {
...
 		anicopy.Duration = math.Abs(step) * ani.Duration
 		p.goAnimate(animname, &anicopy)
 		return
 	}
-	p.goMoveForward(step * float64(p.g.stepUnit))
+	p.goMoveForward(step * float64(p.g.gridUnit))
```
Comment (xushiwei) [inline sprite.go:753]: `step` and `move` both use `world axis`. so it should be `p.goMoveForward(step)`.

File: game.go [hunk 10]
```diff
@@ -498,10 +500,11 @@ func (p *Game) loadIndex(g reflect.Value, index interface{}) (err error) {
 	} else {
 		p.baseObj.initWithSize(proj.Map.Width, proj.Map.Height)
 	}
```
Comment (xushiwei) [inline game.go:502]: value of `worldWidth_/worldHeight_`:
* if user specifies map.width and map.height, then: worldWidth_, worldHeight_ = map.width, map.height.
* if not, then: worldWidth_, worldHeight_ = float64(sceneImg.width) / map.gridUnit, float64(sceneImg.height) / map.gridUnit.

File: game.go [hunk 11]
```diff
@@ -311,12 +312,13 @@ type cameraConfig struct {
...
-	Width  int    `json:"width"`
-	Height int    `json:"height"`
+	Width    int     `json:"width"`
+	Height   int     `json:"height"`
+	Mode     string  `json:"mode"`
+	GridUnit float64 `json:"gridUnit"`
 }
 
-func toMapMode(mode string) int {
+func tomapMode(mode string) int {
```
Comment (xushiwei) [inline game.go:321]: `toMapMode` is better.

File: game.go (hunk 10 above)
Comment (xushiwei) [reply game.go:502]: value of `windowWidth_/windowHeight_`:

* width, height := int(worldWidth_ * map.gridUnit), int(worldHeight_ * map.gridUnit)
* if user specifies conf.width and conf.height, then: windowWidth_, windowHeight = conf.width, conf.height (if larger then width/height, limit it).
* if not, then: windowWidth_, windowHeight = width, height (if out of screen size, limit it).

File: sprite.go (hunk 8 above)
Comment (sunqirui1987) [reply sprite.go:748]: this time compatible first,  the next pr change

File: sprite.go (hunk 9 above)
Comment (sunqirui1987) [reply sprite.go:753]: this time compatible first, the next pr change

File: game.go (hunk 11 above)
Comment (sunqirui1987) [reply game.go]: fixed

File: game.go [hunk 12]
```diff
@@ -496,12 +498,19 @@ func (p *Game) loadIndex(g reflect.Value, index interface{}) (err error) {
 	if scenes := proj.getScenes(); len(scenes) > 0 {
 		p.baseObj.init("", scenes, proj.getSceneIndex())
 	} else {
+		if p.windowWidth_ > proj.Map.Width {
+			p.worldWidth_ = p.windowWidth_
```
Comment (xushiwei) [inline game.go:502]: why change world size?
if window is too big, limit window not world.

Comment (sunqirui1987) [reply game.go:502]: if window is too big,  
-----------
 yes!

, limit window not world.
-----
ok

Comment (sunqirui1987) [reply game.go]: fixed

**Pull request merged**